src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs

```diff
@@ -89,40 +89,24 @@ public void Release(object obj)
     /// <summary>
     /// Used to cache a weak reference in a threadsafe way
     /// </summary>
-    internal sealed class SharedReference
+    internal sealed class SharedReference<T> where T : class
     {
-        private readonly WeakReference _ref = new WeakReference(null);
-        private volatile int _locked;
+        private readonly WeakReference<T> _ref = new WeakReference<T>(null);
 
-        /// <summary>
-        /// Return an object from a weakref, protected by a lock.
-        /// 
-        /// If the exclusive lock can't be obtained, null is returned;
-        /// Note that _ref.Target is referenced only under the protection of the lock. (Is this necessary?)
-        /// </summary>
-        public object Get()
+        public T Get()
         {
-            if (0 == Interlocked.Exchange(ref _locked, 1))
+            lock (this)
```

Comment:
WeakReference is thread-safe. This lock is useless.

Reply:
> Any public static ( Shared in Visual Basic) members of this type are thread safe. Any instance members are not guaranteed to be thread safe.

https://msdn.microsoft.com/en-us/library/gg712738(v=vs.110).aspx

Reply:
This is a boiler text that used to be added everywhere by default. It is wrong in this case.

Reply:
thanks!

Reply:
cc @mairaw